certora/specs/CER-2-Operator/CER-52-Operator-deauthorization.spec

```diff
@@ -0,0 +1,48 @@
+import "../utils/IsMustRevertFunction.spec";
+import "CER-68-Operator-authorization.spec";
+
+methods {
+    function getOwnerOf(bytes19) external returns (address) envfree;
+    function getOperator(bytes19, address) external returns (uint256) envfree;
+    function getAddressPrefix(address) external returns (bytes19) envfree;
+    function haveCommonOwner(address account, address otherAccount) external returns (bool) envfree;
+}
+
+// CER-52 Account Operator that is authorized to operate on behalf the Account MUST only be allowed to be deauthorized by:
+// - the Account Owner, or
+// - the Account Operator itself (one Account Operator MUST NOT be able to deauthorize the other Account Operator)
+
+rule operatorDeauthorizationSetOperator {
+    env e;
+
+    bytes19 addressPrefix;
+    address operator;
+    uint256 operatorBitField;
+
+    address caller = actualCaller(e);
+
+    // call the setOperator() method giving 0 as the bit field to deauthorize
+    setOperator(e, addressPrefix, operator, 0);
+    // since the function did not revert the caller must be
+    // the owner or the operator mentioned
+    assert caller == getOwnerOf(addressPrefix) || caller == operator;
```

Comment:
`setOperator` should only succeed if called by an owner. we even proved it here:
`certora/specs/CER-2-Operator/CER-68-Operator-authorization.spec`

if `setOperator` were to allow being called by the operator to deauthorize itself, it would have to parse provided `operatorBitField` bit by bit.

shall we assert that caller is not operator instead?

Reply:
You are right and that is clearly a stronger rule. It is fixed to work the way you suggested.